### ISSUE TITLE
no debug output

### ISSUE THREAD

**Issue by dschneiderch** [danforthcenter/plantcv#453]
I think I'm still generally confused by the workflow pipelines.

I'm following your [VIS Tutorial](https://plantcv.readthedocs.io/en/stable/vis_tutorial/) with my own image.
https://emailwsu-my.sharepoint.com/:i:/g/personal/dominik_schneider_wsu_edu/EfTahEzq8jpEsXkPB9ZaOiwBDBHwefipD9W1LxtXDhF6pA?e=c09Ty2

I'm running this at the commandline:

```
python vis_workflow.py -i "data\D7-chickpea_test_slowgrowth-2019-09-27 04_16_23-VIS0.png"  -o debug-images -r results.txt -w -D 'print'
```

This should produce a results.txt (True) but also print debug images to debug-images (False).... correct?
Running this command does not produce any errors. I also tried renamiing the image to have no delimiters but it didn't change anything.
Why does the `-w` flag not accept True or False?

Also, I won't be able to run over the entire directory of .png images until the date thing is sorted, correct?

Windows, Conda, plantcv 3.6 via bioconda
 

```
# !/usr/bin/python
import sys, traceback
import cv2
import numpy as np
import argparse
import string
from plantcv import plantcv as pcv
import os
import glob
import numpy as np
from matplotlib import pyplot as plt


### Parse command-line arguments
def options():
    parser = argparse.ArgumentParser(
        description="Imaging processing with opencv")
    parser.add_argument("-i",
                        "--image",
                        help="Input image file.",
                        required=True)
    parser.add_argument("-o",
                        "--outdir",
                        help="Output directory for image files.",
                        required=False)
    parser.add_argument("-r", "--result", help="result file.", required=False)
    parser.add_argument("-w",
                        "--writeimg",
                        help="write out images.",
                        default=False,
                        action="store_true")
    parser.add_argument(
        "-D",
        "--debug",
        help=
        "can be set to 'print' or None (or 'plot' if in jupyter) prints intermediate images.",
        default=None)
    args = parser.parse_args()
    return args


#### Start of the Main/Customizable portion of the workflow.


### Main workflow
def main():
    # Get options
    args = options()

    pcv.params.debug = args.debug  # set debug mode
    pcv.params.debug_outdir = args.outdir  # set output directory

    # Read image
    img, path, filename = pcv.readimage(filename=args.image)

    img_wb = pcv.white_balance(img, mode = 'max', roi = (1455,3085,100,100))
    # imgcheck = np.all(img_wb==255, axis=2)
    # imgcheck1 = np.all(img==255, axis=2)
    # pcv.image_subtract(imgcheck,imgcheck1)

    a_img = pcv.rgb2gray_lab(img_wb, 'a')
    he_a_img=pcv.hist_equalization(a_img)
    thresh = pcv.threshold.binary(he_a_img,20, 255,'dark')
    blur_bin = pcv.median_blur(thresh, 5)
    fill_blur_bin = pcv.fill(blur_bin, 15)

    img_masked = pcv.apply_mask(img_wb, fill_blur_bin, 'black')

    # Identify objects
    id_objects, obj_hierarchy = pcv.find_objects(img=img_masked, mask=fill_blur_bin)

    # Define ROI
    roi1, roi_hierarchy = pcv.roi.rectangle(img=img_masked,
                                            x=500,
                                            y=600,
                                            h=2300,
                                            w=1800)

    # Decide which objects to keep
    roi_objects, hierarchy3, kept_mask, obj_area = pcv.roi_objects(
        img=img,
        roi_contour=roi1,
        roi_hierarchy=roi_hierarchy,
        object_contour=id_objects,
        obj_hierarchy=obj_hierarchy,
        roi_type='partial')

    # Object combine kept objects
    obj, mask = pcv.object_composition(img=img,
                                    contours=roi_objects,
                                    hierarchy=hierarchy3)

    ############### Analysis ################

    outfile = True
    if args.writeimg == True:
        outfile = args.outdir + "/" + filename

    # Determine color properties: Histograms, Color Slices, output color analyzed histogram (optional)
    color_histogram = pcv.analyze_color(rgb_img=img,
                                        mask=kept_mask,
                                        hist_plot_type='all')

    # Pseudocolor the grayscale image
    pseudocolored_img = pcv.visualize.pseudocolor(gray_img=a_img,
                                                  mask=kept_mask,
                                                  cmap='viridis')

    # Write shape and color data to results file
    pcv.print_results(filename=args.result)


if __name__ == '__main__':
    main()
```

**Comment by nfahlgren** [comment]
Hi @dschneiderch,

I was able to reproduce this on Windows. It looks like the issue is that the argument `print` cannot be quoted on Windows. Linux and macOS strip off the quotes but Windows appears to be doing something else. It doesn't need to be quoted though, so `-D print` will work.

The `-w` option is a flag-type argument, so it doesn't accept True/False because it automatically has a boolean value. When the flag is absent it is `False` by default, and when it is present `action="store_true"` sets the value to `True`. But another note about `-w`, in this script it has no effect. It was intended as a way to turn result-type output images on/off (as opposed to the debug images) but the flag requires the user to build in the logic for saving the images or not, at least at present. 

But that being said, for your purposes it doesn't necessarily matter since you are saving the debug images.

I need to think on a short-term quick solution for the date problem. I have some ideas but I keep thinking they are more involved than I'm anticipating...

**Comment by nfahlgren** [comment]
This would be sort of an advanced option, but I think I could add an option to input a regular expression. I'm thinking something like this would work, using the filename `D7-chickpea_test_slowgrowth-2019-09-27 04_16_23-VIS0.png` as an example:

Inputs (note I just guessed on the metadata terms):
```
--meta plantbarcode_measurementlabel_timestamp_camera
--delimiter _
--regex '(.+)-(.+)-(\d{4}-\d{2}-\d{2} \d{2}_\d{2}_\d{2})-(.+)'
```

The meta input is the filename metadata structure separated by the input delimiter (default = _), but the image is not actually named with underscores. But I can still process the metadata terms by splitting `meta` on `delimiter`. Then since `regex` is defined, I can process image names using the regex instead of `meta` and `delimiter`.

This example regex looks for:

1. any character, 1 or more
2. dash
3. any character, 1 or more
4. dash
5. 4 digits, dash, 2 digits,  dash, 2 digits, space, 2 digits, underscore, 2 digits, underscore, 2 digits
6. dash
7. any character, 1 or more

The parenthesis mark where we want to store the match, which returns the following groups:

```
('D7', 'chickpea_test_slowgrowth', '2019-09-27 04_16_23', 'VIS0')
``` 

Then these would be matched to the metadata:

```
plantbarcode = D7
measurementlabel = chickpea_test_slowgrowth
timestamp = 2019-09-27 04_16_23
camera = VIS0
```

Regex syntax can be kind of crazy, but what do you think?

**Comment by dschneiderch** [comment]
Crazy regex! Thanks, looks perfect! It's always a bit of mystery to me and involves much flailing until i get lucky..

So the `--delim _` is irrelevant when `--regex` is present?

Can the timestamp be parsed that way? i assume python has a strptime() that can parse any custom format.

I think having a generic regex syntax for metadata would be powerful. even if it is a bit complicated. you might consider making --delim and --regex mutually exclusive so it's clear what the user intends.

Alternatively, I can write a small script that preprocesses the filenames with that regex...

**Comment by dschneiderch** [comment]
Also, so for `-w`  you are expecting something like this?

```
if w:
   pcv.pseudocolor.visualize(...)
```

**Comment by nfahlgren** [comment]
I'm thinking the delimiter would be irrelevant for parsing the filename when regex is present, but still relevant for encoding the meta variables. With the regex we can parse the filename into components but in order to know what each component refers to, we would still need the `--meta` input. So I was thinking the metadata terms could be encoded with the delimiter, e.g.:

`plantbarcode_measurementlabel_timestamp_camera`

As long as the user uses the default delimiter they don't need to set `--delimiter`.

Although alternatively, we could change `--meta` to accept a comma-separated list of terms in all cases and only use `--delimiter` or `--regex` as mutually exclusive options, which sounds simpler as I'm thinking about this as I type :)

**Comment by nfahlgren** [comment]
That's correct, `args.writeimg` or `args.w` can be used in an if statement to control user-defined output images. It can be ignored if you want to always have those outputs on or off (absent).

**Comment by dschneiderch** [comment]
> 
> 
> That's correct, `args.writeimg` or `args.w` can be used in an if statement to control user-defined output images. It can be ignored if you want to always have those outputs on or off (absent).

actually `args.w` does not work. only `args.writeimg`

Its a bit confusing that 'print' may not print something. Also, I also find the lack of distinction between a directory for debug and a directory for potential output to be confusing. Easy enough to program around but it'd be nice to be able to specify a separate output directory for images (that are not from debug=print).  Personally I would expected `--debugdir` and `-outputdir` . When I ready to roll over all images, i'll want debug="None" but I might still want the histogram and pseudocolor as an output. Timelapse movies of the pseudocolor images are very popular here.

Also, would it be dangerous to make sure the path specified with -o exists so the user doesn't hbave to do `os.makedirs(outfile, exist_ok=True)` ?

```
    if args.writeimg == True:
                outfile = args.outdir + "/" + os.path.splitext(filename)[0]
                os.makedirs(outfile, exist_ok=True)

        # Determine color properties: Histograms, Color Slices, output color analyzed histogram (optional)
        color_histogram = pcv.analyze_color(rgb_img=img,
                                            mask=kept_mask,
                                            hist_plot_type='all')

    # Pseudocolor the grayscale image
        pseudocolored_img = pcv.visualize.pseudocolor(gray_img=a_img,
                                                  mask=kept_mask,
                                                  cmap='viridis')

    pseudocolored_img.savefig(os.path.join(outfile,'falsecolor.png'))
```

**Comment by nfahlgren** [comment]
The output directory for debug images need not be the same as `--outdir`.

The `options` function needs to have a few specific flags to be compatible with `plantcv-workflow.py`, including `--outdir`. Out intention for `outdir` was that it's where you want to store the analysis output images. You can add other flags to your workflow though, so you could add a custom `--debugdir` flag that is used to set `pcv.params.debug_outdir`. To set it via `plantcv-workflow.py` you can use the `--other_args` option:

`--other_args '--debugdir /path/to/debugdir'`

**Comment by dschneiderch** [comment]
Great. Let me know if you have a branch you want me to test.
________________________________

**Comment by nfahlgren** [comment]
Hi @dschneiderch,

I think the regex parser is working now but could use some broader testing (branch: add-regex-metadata-parser). I did not end up adding a `--regex` option like I thought I was going to but instead set it up so the existing `--delimiter` option will accept a regex pattern instead of a single delimiter character. The other change is that the `--meta` option now always needs a comma-separated list of terms, not terms separated by the `--delimiter`. So the key new options for `plantcv-workflow.py` would look like this now for your data:

```
--meta plantbarcode,measurementlabel,timestamp,camera
--delimiter '(.+)-(.+)-(\d{4}-\d{2}-\d{2} \d{2}_\d{2}_\d{2})-(.+)'
```

Note, per #457 I did not change any of the flags in my pull request to keep things simple for now.
